src/compiler/scala/tools/nsc/transform/patmat/MatchAnalysis.scala

```diff
@@ -164,6 +164,8 @@ trait TreeAndTypeAnalysis extends Debugging {
         case sym if sym.isCase =>
           List(List(tp))
 
+        case sym if sym.isTypeSkolem => enumerateSubtypes(sym.info.upperBound, grouped) // pos/t12277
```

Comment:
One concern I was just reminding myself of is the fact that for some `Type`s `upperBound` returns `this`.  I don't _think_ such types belong to a `TypeSkolem` symbol (or, more generally, a symbol which `isTypeSkolem`) but it's a possibility still, maybe???

In trying to induce the infinite loop I'm concerned about it turns out not to happen:

```scala
scala> def matcher[A](x: A) = x match { case null => 1 }
                              ^
       warning: match may not be exhaustive.
```

That warning, btw, exists without my patch here...